### PR TITLE
[test] [ir] Add tests for the alias_analysis pass

### PR DESCRIPTION
**Pull request by k-ye**
Related issue = #2193 

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----

### CODE REVIEW

File: tests/cpp_new/analysis/alias_analysis_test.cpp [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+  for (int i = 0; i < 2; ++i) {
+    snode->physical_index_position[i] = i;
+  }
+  snode->extractors[snode->physical_index_position[2]].trailing_bits = 2;
+
+  IRBuilder builder;
+  // The difference along the 2-dimension is (6 -3) = 3, which is smaller than
+  // (1 << trailing_bits) = 4.
+  const auto indices1 = make_const_indices({1, 2, 3}, &builder);
+  const auto indices2 = make_const_indices({1, 2, 6}, &builder);
```
Comment (xumingkuan) [inline tests/cpp_new/analysis/alias_analysis_test.cpp:77]: Well... IIUC `a[1, 2, 3]` and `a[1, 2, 6]` are **different** elements when `trailing_bits == 2`.

Maybe we should change
```C++
      if (!diff.first) {
        uncertain = true;
      } else if (std::abs(diff.second) >=
                 (1 << snode->extractors[snode->physical_index_position[i]]
                           .trailing_bits)) {
        return AliasResult::different;
      }
```
to
```C++
      if (!diff.first || diff.second != 0) {
        uncertain = true;
      }
      if (std::abs(diff.second) >=
          (1 << snode->extractors[snode->physical_index_position[i]]
                           .trailing_bits)) {
        return AliasResult::different;
      }
```

**Review by xumingkuan — COMMENTED** [review]
Thanks!

File: tests/cpp_new/analysis/alias_analysis_test.cpp (hunk 1 above)
Comment (k-ye) [reply tests/cpp_new/analysis/alias_analysis_test.cpp:77]: Hmm, could you explain the original intention when writing this in https://github.com/taichi-dev/taichi/pull/1657?

**Comment by xumingkuan** [comment]
Well... For `trailing_bits == 2`, indices `0, 1, 2, 3` point to the same address, and so do `4, 5, 6, 7`, `8, 9, 10, 11`, etc.


So when two indices are different by >= 4, we are sure that they point to the different address.


However, I forgot to add the condition that if two indices are different by < 4, we are not sure if they point to the same address...







在 2021-04-06 08:44:04，"Ye Kuang" ***@***.***> 写道：

@k-ye commented on this pull request.

In tests/cpp_new/analysis/alias_analysis_test.cpp:

> +  // (1 << trailing_bits) = 4.
+  const auto indices1 = make_const_indices({1, 2, 3}, &builder);
+  const auto indices2 = make_const_indices({1, 2, 6}, &builder);


Hmm, could you explain the original intention when writing this in #1657?

File: tests/cpp_new/analysis/alias_analysis_test.cpp (hunk 1 above)
Comment (k-ye) [reply tests/cpp_new/analysis/alias_analysis_test.cpp:77]: Thanks, fixed

**Review by xumingkuan — APPROVED** [review]
LGTM, thanks!

**Pull request merged**